### PR TITLE
Support for TLS client settings for clustering

### PR DESCRIPTION
**Pull request by tiagorossig**
### PR Description

Clustering in TLS environments was not possible before these changes.

#### Why is this change required
My organization runs alloy pods in environments that require all pod communication to be over HTTPS with TLS encryption. When enabling clustering, we noticed that communication between pods was blocked because it was going over HTTP.

#### Testing
Built `alloy` from this branch and used the `enable-tls` CLI flag with the appropriate certificate material to configure the TLS client. Made sure that pods can still communicate amongst themselves and constantly converge on the cluster's state.

Some logs that confirm pod communication still works:
```
{"ts":"2024-09-20T22:36:54.802881244Z","level":"debug","service":"cluster","component":"memberlist","ts":"2024/09/20 22:36:54","caller":"[DEBUG] memberlist: Initiating push/pull sync with: <ip>.ec2.internal <pod1-ip>"}
{"ts":"2024-09-20T22:36:54.804031852Z","level":"debug","msg":"merging remote state","service":"cluster","remote_time":34}
{"ts":"2024-09-20T22:37:02.656126162Z","level":"debug","msg":"received DNS query response","service":"cluster","addr":"<cluster-address>","record_type":"A/AAAA","records_count":3}
{"ts":"2024-09-20T22:37:02.656152001Z","level":"debug","msg":"discovered peers","service":"cluster","peers_count":3,"peers":"<pod0-ip>,<pod1-ip>,<pod2-ip>"}
{"ts":"2024-09-20T22:37:02.656158483Z","level":"info","msg":"rejoining peers","service":"cluster","peers_count":3,"peers":"<pod0-ip>,<pod1-ip>,<pod2-ip>"}
{"ts":"2024-09-20T22:37:02.657273362Z","level":"debug","service":"cluster","component":"memberlist","ts":"2024/09/20 22:37:02","caller":"[DEBUG] memberlist: Initiating push/pull sync with:  <pod0-ip>"}
{"ts":"2024-09-20T22:37:02.658438004Z","level":"debug","msg":"merging remote state","service":"cluster","remote_time":35}
{"ts":"2024-09-20T22:37:02.65934668Z","level":"debug","service":"cluster","component":"memberlist","ts":"2024/09/20 22:37:02","caller":"[DEBUG] memberlist: Initiating push/pull sync with:  <pod1-ip>"}
{"ts":"2024-09-20T22:37:02.660418062Z","level":"debug","msg":"merging remote state","service":"cluster","remote_time":34}
{"ts":"2024-09-20T22:37:02.660630938Z","level":"debug","service":"cluster","component":"memberlist","ts":"2024/09/20 22:37:02","caller":"[DEBUG] memberlist: Stream connection from=<pod2-ip>"}
{"ts":"2024-09-20T22:37:02.660723496Z","level":"debug","service":"cluster","component":"memberlist","ts":"2024/09/20 22:37:02","caller":"[DEBUG] memberlist: Initiating push/pull sync with:  <pod2-ip>"}
{"ts":"2024-09-20T22:37:02.661081355Z","level":"debug","msg":"merging remote state","service":"cluster","remote_time":36}
```

Grafana Alloy debugging UI also shows that clustered pods can still recognize each other:
Screenshot 2024-09-20 at 5 49 24 PM

### Which issue(s) this PR fixes

Fixes https://github.com/grafana/alloy/issues/367

### Notes to the Reviewer

This PR depends on https://github.com/grafana/ckit/pull/63, which has not yet been merged.

### PR Checklist

- [x] CHANGELOG.md updated
- [x] Documentation added
- [ ] Tests updated
- [ ] Config converters updated

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -917,3 +917,6 @@ replace github.com/prometheus/procfs => github.com/prometheus/procfs v0.12.0
 // It's important to remove it asap because in version v0.13.1 there is a fix for Beyla.
 // PR to track it: https://github.com/opencontainers/runc/pull/4397
 replace github.com/opencontainers/runc => github.com/rafaelroquetto/runc v1.1.14-1
+
+// Temporary replace until ckit changes are merged upstream
+replace github.com/grafana/ckit => github.com/tiagorossig/ckit v0.0.0-20240920184404-077657c65a6f
```
Comment (tiagorossig) [inline go.mod:922]: Removal of replace directive depends on merging https://github.com/grafana/ckit/pull/63

File: docs/sources/reference/cli/run.md [hunk 2]
```diff
@@ -112,6 +117,9 @@ The comma-separated list of addresses provided in `--cluster.join-addresses` can
 In both cases, the port number can be specified with a `:<port>` suffix. If ports are not provided, default of the port used for the HTTP listener is used.
 If you do not provide the port number explicitly, you must ensure that all instances use the same port for the HTTP listener.
 
+The `--cluster.enable-tls` flag can be set to enable TLS for peer-to-peer communications. Additional arguments are required to configure the TLS client, including the CA certificate (`--cluster.tls-ca-path`), 
+the certificate (`--cluster.tls-cert-path`), the key (`--cluster.tls-key-path`), and the server name (`--cluster.tls-server-name`).
```
[the remark this answers is not in the document]
Comment (tiagorossig) [reply docs/sources/reference/cli/run.md]: ✅

File: go.mod (hunk 1 above)
Comment (tiagorossig) [reply go.mod]: Done https://github.com/grafana/alloy/pull/1724/commits/ead8f7ca1d52e2d8dbccbb75c52741bd37cf5271

**Review by mattdurham — APPROVED** [review]
LGTM

**Comment by mattdurham** [comment]
Looks good to me, if you can resolve the conflict automerge will merge it.

**Review by clayton-cornell — APPROVED** [review]

**Comment by tiagorossig** [comment]
@mattdurham Much appreciated! Resolved the `CHANGELOG.md` conflict

**Comment by tiagorossig** [comment]
Edited to pass CI https://github.com/grafana/alloy/actions/runs/11131417334/job/30934202706?pr=1724

File: internal/service/cluster/cluster.go [hunk 3]
```diff
@@ -121,26 +127,35 @@ func New(opts Options) (*Service, error) {
...
-				// TODO(rfratto): consider making the max timeout configurable.
-				timeout := 30 * time.Second
-				if dur, ok := deadlineDuration(ctx); ok && dur < timeout {
-					timeout = dur
-				}
-
-				return net.DialTimeout(network, addr, timeout)
-			},
+	httpTransport := &http2.Transport{
+		AllowHTTP: false,
```
Comment (thampiotr) [inline internal/service/cluster/cluster.go:134]: @wildum perhaps this caused the outage we're seeing?

Comment (wildum) [reply internal/service/cluster/cluster.go:134]: according to https://github.com/grafana/ckit/blob/main/node.go#L141, AllowHTTP should be set to true for non-TLS connections

**Pull request merged**